File: 12-Heaps/cpp/06-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        for (int num : nums) {
            freq[num]++;
        }

        vector<vector<int>> bucket(nums.size() + 1);
        for (const auto& [num, count] : freq) {
            bucket[count].push_back(num);
        }

        vector<int> res;
        for (int i = bucket.size() - 1; i >= 0 && (int)res.size() < k; i--) {
            for (int num : bucket[i]) {
                res.push_back(num);
                if ((int)res.size() == k) break;
            }
        }

        return res;
    }
};

File: 12-Heaps/java/06-TopKFrequentElements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();

        for (int num : nums) {
            freq.put(num, freq.getOrDefault(num, 0) + 1);
        }

        List<Integer>[] bucket = new List[nums.length + 1];

        for (int num : freq.keySet()) {
            int f = freq.get(num);

            if (bucket[f] == null) {
                bucket[f] = new ArrayList<>();
            }

            bucket[f].add(num);
        }

        int[] res = new int[k];
        int idx = 0;

        for (int i = bucket.length - 1; i >= 0 && idx < k; i--) {
            if (bucket[i] != null) {
                for (int num : bucket[i]) {
                    res[idx++] = num;
                    if (idx == k) {
                        break;
                    }
                }
            }
        }

        return res;
    }
}
